### PR TITLE
Graduate `SeccompDefault` feature to stable / GA

### PR DESCRIPTION
**Pull request by saschagrunert**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Updating the required code and docs for `SeccompDefault` to go GA, which now means that we enable the feature per default.

#### Which issue(s) this PR fixes:

Refers to https://github.com/kubernetes/enhancements/issues/2413

#### Special notes for your reviewer:

cc @kubernetes/sig-node-api-reviews

#### Does this PR introduce a user-facing change?

```release-note
Graduated seccomp profile defaulting to GA.

Set the kubelet `--seccomp-default` flag or `seccompDefault` kubelet configuration field to `true` to make pods on that node default to using the `RuntimeDefault` seccomp profile.

Enabling seccomp for your workload can have a negative performance impact depending on the kernel and container runtime version in use.

Guidance for identifying and mitigating those issues is outlined in the Kubernetes [seccomp tutorial](https://k8s.io/docs/tutorials/security/seccomp).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
KEP: https://github.com/kubernetes/enhancements/tree/05f81b8/keps/sig-node/2413-seccomp-by-default
```

### CODE REVIEW

File: test/e2e/node/security_context.go
Comment (bart0sh) [inline test/e2e/node/security_context.go:221]: > we now use `RuntimeDefault` instead of `Unconfined`

Should `scenarioName` parameter be set to "seccomp default RuntimeDefault"?

Comment (saschagrunert) [reply test/e2e/node/security_context.go:221]: Exactly, fixed as suggested :+1:

**Comment by dims** [comment]
@saschagrunert does `kubelet --help` prints the right default?

**Comment by saschagrunert** [comment]
> @saschagrunert does `kubelet --help` prints the right default?

Yep:

```
> make WHAT=cmd/kubelet
> ./_output/bin/kubelet -h | rg Seccomp
      --seccomp-default RuntimeDefault                           Enable the use of RuntimeDefault as the default seccomp profile for all workloads. The SeccompDefault feature gate must be enabled to allow this flag. (default true)
```

**Comment by dims** [comment]
/lgtm

thanks @saschagrunert

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [inline pkg/features/kube_features.go:1021]: should be locked to default as per https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-stages

Comment (SergeyKanzhelev) [reply pkg/features/kube_features.go:1021]:
```suggestion
	SeccompDefault: {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
```

File: cmd/kubelet/app/options/options.go
Comment (SergeyKanzhelev) [inline cmd/kubelet/app/options/options.go:315]:
```suggestion
	fs.BoolVar(&f.SeccompDefault, "seccomp-default", f.SeccompDefault, "Enable the use of `RuntimeDefault` as the default seccomp profile for all workloads.")
```

File: pkg/features/kube_features.go
Comment (SergeyKanzhelev) [reply pkg/features/kube_features.go:1021]: pease also remove all the usage of this feature gate in code

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:767]:
```suggestion
```

File: test/e2e/node/security_context.go
Comment (SergeyKanzhelev) [inline test/e2e/node/security_context.go:218]: unrelated to this PR: this whole file has `[LinuxOnly]` tests. Should it be excluded for Windows compilation?

Comment (saschagrunert) [reply test/e2e/node/security_context.go:220]: I think we can, let me give it a try.

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

**Comment by liggitt** [comment]
> Action required: SeccompDefault graduated to GA, which means that the corresponding `--seccomp-default`
> kubelet CLI flag as well as configuration option `seccompDefault` now default to `true`.

That does not match what was previously discussed in https://github.com/kubernetes/kubernetes/pull/101943#issuecomment-849981663

The feature can graduate to GA but the flag should remain opt-in.

File: pkg/features/kube_features.go
Comment (liggitt) [inline pkg/features/kube_features.go:1021]: add a comment to remove this gate in 1.29

File: test/e2e/storage/utils/utils.go
Comment (liggitt) [inline test/e2e/storage/utils/utils.go:388]: I'm surprised by this change... is this an example of the kind of workloads making `--seccomp-default` default to `true` would break on upgrade?

File: cmd/kubelet/app/options/options.go
Comment (SergeyKanzhelev) [reply cmd/kubelet/app/options/options.go:154]: there seems that was another discussion: https://github.com/kubernetes/enhancements/pull/3718#issuecomment-1416526973 This is what I used as a basis for lgtm. I definitely remember beta stage discussion

Comment (liggitt) [reply cmd/kubelet/app/options/options.go:154]: that comment references the performance impact, it doesn't address the "workloads could arbitrarily stop working on upgrade" issue, which is a *much* bigger concern

Comment (SergeyKanzhelev) [reply cmd/kubelet/app/options/options.go:154]: I understand, the result of this PR is that GA requirements in KEP's readme are saying that seccomp will be ON by default though.

File: test/e2e/storage/utils/utils.go
Comment (saschagrunert) [reply test/e2e/storage/utils/utils.go:388]: Yes, at least with containerd v1.6.9 being used in CI. I tested the same in CRI-O and the test succeeded. I assume it's due to slight differences in the `RuntimeDefault` profile.

File: cmd/kubelet/app/options/options.go
Comment (saschagrunert) [reply cmd/kubelet/app/options/options.go]: Ah, yeah now I can remember as well, that's a bummer. cc @mrunalp 

I'm thinking if a corresponding API Server feature (after the GA graduation of the kubelet feature) would help us here, but that's another story.

File: test/e2e/node/security_context.go
Comment (liggitt) [inline test/e2e/node/security_context.go:1]: this seems weird... the test is marked as `[LinuxOnly]` but can build/run from any os (the test runs as an API client)

File: pkg/features/kube_features.go
Comment (liggitt) [inline pkg/features/kube_features.go:1021]: sweep and remove checks of `...Enabled(features.SeccompDefault)` now that this is locked true

File: test/e2e/node/security_context.go
Comment (saschagrunert) [reply test/e2e/node/security_context.go]: Makes sense, I removed the build constraint.

File: pkg/features/kube_features.go
Comment (saschagrunert) [reply pkg/features/kube_features.go:1021]: Ah yeah makes sense, I removed the checks.

**Comment by liggitt** [comment]
/approve
for kubelet config API changes

will leave lgtm to node review

**Comment by dims** [comment]
this PR by itself LGTM. folks have pointed out that the KEP text needs to match this PR. I am ok either to land this first and then update the KEP or the other way around.

**Comment by SergeyKanzhelev** [comment]
@saschagrunert would you please also address the todo comment that is relatively related: https://github.com/kubernetes/kubernetes/blob/e55f2a9b54e0f6bd9ed12e50752d7d2545ab0cab/pkg/kubelet/kuberuntime/kuberuntime_sandbox.go#L171-L174

this PR is lgtm

/lgtm
/hold

(hold if you will decide to address TODO as part of this PR, feel free to unhold any moment)

**Comment by saschagrunert** [comment]
@SergeyKanzhelev I'm going to follow-up on the `TODO`, thank you for the hint!

/unhold

**Comment by sftim** [comment]
My suggestion for a changelog entry (we can revise these after merge):
```markdown
Graduated seccomp profile defaulting to GA.

Set the kubelet `--seccomp-default` flag or `seccompDefault` kubelet configuration field to `true` to make
pods on that node default to using the `RuntimeDefault` seccomp profile.

Enabling seccomp for your workload can have a negative performance impact depending on the kernel and
container runtime version in use.
Guidance for identifying and mitigating those issues is outlined in the Kubernetes
[seccomp tutorial](https://k8s.io/docs/tutorials/security/seccomp).
```

---

Should we also track work to update related docs, such as https://kubernetes.io/docs/concepts/security/security-checklist/?

**Comment by saschagrunert** [comment]
> My suggestion for a changelog entry (we can revise these after merge):
> 
> ```gfm
> Graduated seccomp profile defaulting to GA.
> 
> Set the kubelet `--seccomp-default` flag or `seccompDefault` kubelet configuration field to `true` to make
> pods on that node default to using the `RuntimeDefault` seccomp profile.
> 
> Enabling seccomp for your workload can have a negative performance impact depending on the kernel and
> container runtime version in use.
> Guidance for identifying and mitigating those issues is outlined in the Kubernetes
> [seccomp tutorial](https://k8s.io/docs/tutorials/security/seccomp).
> ```
> 
> Should we also track work to update related docs, such as https://kubernetes.io/docs/concepts/security/security-checklist/?

Thanks, changed the release notes as suggested :+1: 

Yes, docs updates will be part of the usual enhancement process and I'm following-up on them in the next weeks.

**Pull request merged**
